internal/core/app_status_report.go

```diff
@@ -0,0 +1,340 @@
+	pb "github.com/hashicorp/waypoint/internal/server/gen"
+)
+
+func (a *App) DeploymentStatusReport(
+	ctx context.Context,
+	deployTarget *pb.Deployment,
+) (*pb.StatusReport, *sdk.StatusReport, error) {
+	var evalCtx hcl.EvalContext
+	// Load the deployment variables context
+	if err := a.deployStatusReportEvalContext(ctx, deployTarget, &evalCtx); err != nil {
+		return nil, nil, err
+	}
+
+	// Load variables from deploy
+	hclCtx := evalCtx.NewChild()
+	if _, err := a.deployEvalContext(ctx, hclCtx); err != nil {
+		return nil, nil, err
+	}
+
+	c, err := a.createStatusReporter(ctx, hclCtx, component.PlatformType)
+	if status.Code(err) == codes.Unimplemented {
+		c = nil
+		err = nil
+	}
+	if err != nil {
+		a.logger.Error("error creating component in platform", "error", err)
+		return nil, nil, err
+	}
+
+	a.logger.Debug("starting status report operation")
+	statusReporter, ok := c.Value.(component.Status)
+
+	if !ok || statusReporter.StatusFunc() == nil {
+		a.logger.Debug("component is not a Status or has no StatusFunc()")
+		return nil, nil, nil
+	}
+	defer c.Close()
+
+	return a.statusReport(ctx, "deploy_statusreport", c, deployTarget)
+}
+
+func (a *App) ReleaseStatusReport(
+	ctx context.Context,
+	releaseTarget *pb.Release,
+) (*pb.StatusReport, *sdk.StatusReport, error) {
+	var evalCtx hcl.EvalContext
+	// Load the deployment variables context
+	if err := a.releaseStatusReportEvalContext(ctx, releaseTarget, &evalCtx); err != nil {
+		return nil, nil, err
+	}
+
+	// Load variables from deploy
+	hclCtx := evalCtx.NewChild()
+	if _, err := a.deployEvalContext(ctx, hclCtx); err != nil {
+		return nil, nil, err
+	}
+
+	c, err := a.createStatusReporter(ctx, &evalCtx, component.ReleaseManagerType)
+	if status.Code(err) == codes.Unimplemented {
+		c = nil
+		err = nil
+	}
+	if err != nil {
+		a.logger.Error("error creating component in platform", "error", err)
+		return nil, nil, err
+	}
+
+	a.logger.Debug("starting status report operation")
+	statusReporter, ok := c.Value.(component.Status)
+
+	if !ok || statusReporter.StatusFunc() == nil {
+		a.logger.Debug("component is not a Status or has no StatusFunc()")
+		return nil, nil, nil
+	}
+	defer c.Close()
+
+	return a.statusReport(ctx, "release_statusreport", c, releaseTarget)
+}
+
+// A generic status report func that takes an already setup component and a
+// specific target to execute the report on.
+func (a *App) statusReport(
+	ctx context.Context,
+	loggerName string,
+	component *Component,
+	target interface{},
+) (*pb.StatusReport, *sdk.StatusReport, error) {
+	if loggerName == "" {
+		loggerName = "statusreport"
+	}
+
+	result, msg, err := a.doOperation(ctx, a.logger.Named(loggerName), &statusReportOperation{
+		Component: component,
+		Target:    target,
+	})
+	if err != nil {
+		return nil, nil, err
+	}
+
+	var status *sdk.StatusReport
+	if result != nil {
+		status = result.(*sdk.StatusReport)
+	}
+
+	return msg.(*pb.StatusReport), status, nil
```

Comment:
Probably best to guard this type cast with the checked version and return an error if msg was not a `*pb.StatusReport` in case there is a bad plugin, we don't want it to crash.